File: brute_force.py
from particle import Particle
from geometry import are_neighbors, are_neighbors_periodic


def brute_force_neighbors(particles: list[Particle], rc: float) -> dict[int, list[int]]:
    neighbors = {p.id: [] for p in particles}

    n = len(particles)
    for i in range(n):
        for j in range(i + 1, n):
            p1 = particles[i]
            p2 = particles[j]

            if are_neighbors(p1, p2, rc):
                neighbors[p1.id].append(p2.id)
                neighbors[p2.id].append(p1.id)

    return neighbors


def brute_force_neighbors_periodic(
    particles: list[Particle], rc: float, L: float
) -> dict[int, list[int]]:
    neighbors = {p.id: [] for p in particles}

    n = len(particles)
    for i in range(n):
        for j in range(i + 1, n):
            p1 = particles[i]
            p2 = particles[j]

            if are_neighbors_periodic(p1, p2, rc, L):
                neighbors[p1.id].append(p2.id)
                neighbors[p2.id].append(p1.id)

    return neighbors

File: src/brute_force.c
#include "brute_force.h"
#include "geometry.h"

static void initialize_neighbor_counts(int *neighbor_count, int n) {
    for (int i = 0; i < n; i++) {
        neighbor_count[i] = 0;
    }
}

void brute_force_neighbors(
    Particle *particles,
    int n,
    double rc,
    int **neighbors,
    int *neighbor_count
) {
    initialize_neighbor_counts(neighbor_count, n);

    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (are_neighbors(particles[i], particles[j], rc)) {
                neighbors[i][neighbor_count[i]] = j;
                neighbor_count[i]++;

                neighbors[j][neighbor_count[j]] = i;
                neighbor_count[j]++;
            }
        }
    }
}

void brute_force_neighbors_periodic(
    Particle *particles,
    int n,
    double rc,
    double L,
    int **neighbors,
    int *neighbor_count
) {
    initialize_neighbor_counts(neighbor_count, n);

    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (are_neighbors_periodic(particles[i], particles[j], rc, L)) {
                neighbors[i][neighbor_count[i]] = j;
                neighbor_count[i]++;

                neighbors[j][neighbor_count[j]] = i;
                neighbor_count[j]++;
            }
        }
    }
}